### ISSUE TITLE
experiment setting and results problem

### ISSUE THREAD

**Issue by andyclsr** [zhmzm/FLAME#2]
Hi, it since like something wrong happened to your email
![image](https://user-images.githubusercontent.com/69943589/230133096-30dba5e6-6915-4631-8b8c-664c919820bc.png)

Firstly, thank you for your kindness and opening source your implementation of FLAME. I really learned a lot from it. But I encountered some unexpected questions when I conducted some simple experiments. And below are the details:
(1) It seems like in your test.py the function add_trigger doesn't contain the modification of the DBA situation.
(2) It seems like you ignore the attack_begin implementation even though it is easy to implement.
(3) when I tested on the setting like this , I conducted two settings : first is to attack in the beginning; second is to attack when epoch is larger 100. But the ASR in the second situation is really low, like staying 10% even lower. It seems like attacking in the latter epoch didn't work. I wonder what your experiment setting is like and could you provide some detailed experiment setting? And it would be better if you could kindly update your final code, since it seems like the repository is not complete. 
![image](https://user-images.githubusercontent.com/69943589/230132672-266da3cb-90df-4376-8437-f3bc8ef120a8.png)

(4) Do you consider the model replacement attack? like scaling models etc. It would be better if you could update this part.
(5) I can't really reproduce the results as you mentioned below and I don't know whether you attack from the beginning or in the latter epoch. And when did you record the ASR? Are the ASRs(for example 70.2) you recorded at the end of training which means the backdoor could last in the end?
![image](https://user-images.githubusercontent.com/69943589/230132731-e5808811-385b-4659-8e83-8b5dc94db033.png)

And I try to conduct the experiment in the setting of "CIFAR-10, ResNet18, Badnet, No defense " and attack in the beginning. But the ASR comes from 100, 98.86, 92.26, 30.16, 83.89, 42.57, 47.11, 34.43, 33.70, 28.96... It is quite strange :(

Whatever, thank you for opening source the code! and wish you a great day!

**Comment by zhmzm** [comment]
Hi,
Thanks for your feedback. As it is not the official code, could you please double-check the implementation of the core of FLAME in the function [flame](https://github.com/zhmzm/FLAME/blob/eda5185755c3e8a6869e54f9df347214814a201e/utils/defense.py#L287)? I am not sure if I write the correct code as same as the paper. 
Q1 and Q2 are fixed.
Q3 and Q5, can you write your instruction here? There are lots of factors that affect the performance of FLAME. In my settings, the attack starts at the beginning with only 10% malicious clients, which are always detected by FLAME. The result of the experiment in my setting is shown here. Please overlook the best backdoor task accuracy in the figure, which is not ASR.
![image](https://user-images.githubusercontent.com/53631431/230255792-bda6fd82-6eed-44e3-abd5-622c81e47a20.png)

Q4, scaling attack is not included. Because scaling attack even enlarges the distance, which is easier to be detected by FLAME. I may also implement it later.
Q5, ASR is the average attack success rate in the last 20 rounds (10%) out of 200 rounds in total. For no defense, it is common to observe the phenomenon you reported in the early rounds like the first 10 rounds. Because it is easy to attack the global model in the early stage leading to a high ASR of 100. But at this stage, the accuracy of the model is low. So benign clients force the global model into the main task, which rapidly overwrites the parameters in the backdoor. I think it is meaningless to observe ASR in the early stage where the accuracy of the global model is low. The result of this setting is shown below. Please overlook the best backdoor task accuracy in the figure, which is not ASR.
![image](https://user-images.githubusercontent.com/53631431/230255529-b80423f2-679a-4390-a03b-29f0e2d23c8d.png)

Please let me know if you have any questions about the code. I am glad to discuss the results and details of this paper with you.

**Comment by andyclsr** [comment]
it seems like your email once again has the same problem. :(
could you please leave another email, I think using email to conmunicate is much more convenient.
thanks again!

**Comment by zhmzm** [comment]
Hi,
My email address is updated to semzm@mail.scut.edu.cn. I am glad to discuss this with you in GitHub, which helps more people track the problems.

**Comment by andyclsr** [comment]
Hi Haomin,
It is really kind of you to consider making more people benefit from it. I am also really glad to discuss with you, too!

In respect of DBA problems, I agree with you that many factors may result in this unexpected result and the fact that different experiment settings may come with different results makes sense. Probably, it is not your implementation problem but the parameter setting problem. Later, I will check other repositories to further study this problems. And I will tell you when I make it.

In respect of cifar epoch problem, the DBA paper represent their results in which the ASR got acceptably high in 200th epoch later. And it also need to pre-train for 200 epochs as paper described.
![image](https://user-images.githubusercontent.com/69943589/230396131-0817181f-299c-406c-9483-92e692efcc11.png)

In respect of the parameter setting , your repository doesn't consider the situation that benign clients and malicious clients may use different local epoch and lr.

After checking the DBA, I will also check the RLR, FLTRUST, FLAME. Did you got satisfactory results compared to those paper reported?

Thanks again for your kindness and your valuable time!

best regards & thanks,
Guofu Xie

**Comment by zhmzm** [comment]
Thanks for your verification. In this repo, the settings of the local learning rate and local epoch are the same among benign clients and malicious clients. As said above, a larger local learning rate and extra local epochs cause larger distances, which leads to detection by FLAME. Once FLAME is shown to be effective in Badnet, it should be effective in defending DBA whatever the local learning and local epochs are.
From my results of experiments, I draw the following conclusions: 1) As for RLR, the simple multi-layer CNN can be protected by RLR, while complex networks like ResNet and VGG are not the targeted models in the original paper. 2) As for FLTrust, it provides effective defense in the early stage but it gradually loses its effectiveness when models are close to converging. 3) FLAME and Krum always perform well to detect malicious clients.

**Comment by andyclsr** [comment]
Thanks for explanations!

**Comment by Chad-dio** [comment]
> Hi Haomin, It is really kind of you to consider making more people benefit from it. I am also really glad to discuss with you, too!
> 
> In respect of DBA problems, I agree with you that many factors may result in this unexpected result and the fact that different experiment settings may come with different results makes sense. Probably, it is not your implementation problem but the parameter setting problem. Later, I will check other repositories to further study this problems. And I will tell you when I make it.
> 
> In respect of cifar epoch problem, the DBA paper represent their results in which the ASR got acceptably high in 200th epoch later. And it also need to pre-train for 200 epochs as paper described. ![image](https://user-images.githubusercontent.com/69943589/230396131-0817181f-299c-406c-9483-92e692efcc11.png)
> 
> In respect of the parameter setting , your repository doesn't consider the situation that benign clients and malicious clients may use different local epoch and lr.
> 
> After checking the DBA, I will also check the RLR, FLTRUST, FLAME. Did you got satisfactory results compared to those paper reported?
> 
> Thanks again for your kindness and your valuable time!
> 
> best regards & thanks, Guofu Xie


Hello，andyclsr.I encountered the same issue as you. I'm currently trying to reproduce several defense methods including FLAME, FLTrust, FLDetector, and RLR. However, I've run into some problems—for example, under the RLR setting, ResNet18 fails to learn and the accuracy stays stuck at around 10%. Also, FLDetector seems unable to defend against even the most basic BadNet attack.

Would it be possible for me to ask you some questions about your experience reproducing these methods? Additionally, if you happen to have any reproduction code available, I would be truly grateful if you could share it with me—it would be a great help to my research.

Thank you very much in advance!

**Comment by andyclsr** [comment]
> > Hi Haomin, It is really kind of you to consider making more people benefit from it. I am also really glad to discuss with you, too!
> > In respect of DBA problems, I agree with you that many factors may result in this unexpected result and the fact that different experiment settings may come with different results makes sense. Probably, it is not your implementation problem but the parameter setting problem. Later, I will check other repositories to further study this problems. And I will tell you when I make it.
> > In respect of cifar epoch problem, the DBA paper represent their results in which the ASR got acceptably high in 200th epoch later. And it also need to pre-train for 200 epochs as paper described. ![image](https://user-images.githubusercontent.com/69943589/230396131-0817181f-299c-406c-9483-92e692efcc11.png)
> > In respect of the parameter setting , your repository doesn't consider the situation that benign clients and malicious clients may use different local epoch and lr.
> > After checking the DBA, I will also check the RLR, FLTRUST, FLAME. Did you got satisfactory results compared to those paper reported?
> > Thanks again for your kindness and your valuable time!
> > best regards & thanks, Guofu Xie
> 
> Hello，andyclsr.I encountered the same issue as you. I'm currently trying to reproduce several defense methods including FLAME, FLTrust, FLDetector, and RLR. However, I've run into some problems—for example, under the RLR setting, ResNet18 fails to learn and the accuracy stays stuck at around 10%. Also, FLDetector seems unable to defend against even the most basic BadNet attack.
> 
> Would it be possible for me to ask you some questions about your experience reproducing these methods? Additionally, if you happen to have any reproduction code available, I would be truly grateful if you could share it with me—it would be a great help to my research.
> 
> Thank you very much in advance!

Hi, I haven't focused on backdoor defense or attack in FL for a long time. And as I remembered, I did not try FLDetector, FLtrust, or RLR. And I only conducted on FLAME and FLAME can defense badnet. Also these defense baselines are too old, if you want to conduct experiments, I recommend you to follow the new code repos to start your experiments.
